File: 142-Linked List Cycle II.java
//Time complexity=O(n)
//Space complexity=O(1)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        // if(head.next==null || head==null)
        //     return null;
        //We are doing this using two pointer which are slow and fast, fast moves at 2x speed compared to that of slow pointer.As both the pointer intersect we stop the loop and place the slow pointer to the start and again start moving both pointer but now the spped of both the pointer is the same. So now when they intersect they show the point where the loop starts 
        ListNode slow=head;
        ListNode fast=head;
        boolean flag=false;
        while(fast!=null && fast.next!=null)
        {
            fast=fast.next.next;
            slow=slow.next;
            if(fast==slow)
            {
                flag=true;
                break;
            }
        }
        if(!flag) return null;
        slow=head;
        while(slow!=fast)
        {
            slow=slow.next;
            fast=fast.next;
        }
        return fast;
    }
}

File: 142. Linked List Cycle II.py

#Time complexity: O(N)
#Space complexity: O(1)

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return None
        
        slow = head
        fast = head
        hasCycle = False

        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                hasCycle = True
                break

        if not hasCycle:
            return None

        slow = head
        while fast != slow:
            slow = slow.next
            fast = fast.next
        return slow

            
        